src/app/clusters/service-area-server/service-area-server.cpp

```diff
@@ -0,0 +1,1104 @@
+        uint32_t mapIndex = 0;
+        MapStructureWrapper supportedMap;
+
+        while (mDelegate->GetSupportedMapByIndex(mapIndex++, supportedMap))
+        {
+            ReturnErrorOnFailure(encoder.Encode(supportedMap));
+        }
+        return CHIP_NO_ERROR;
+    });
+}
+
+CHIP_ERROR Instance::ReadSelectedLocations(AttributeValueEncoder & aEncoder)
+{
+    if (mDelegate->GetNumberOfSelectedLocations() == 0)
+    {
+        return aEncoder.EncodeNull();
+    }
+
+    return aEncoder.EncodeList([this](const auto & encoder) -> CHIP_ERROR {
+        uint32_t locationIndex = 0;
+        uint32_t selectedLocation;
+
+        while (mDelegate->GetSelectedLocationByIndex(locationIndex++, selectedLocation))
+        {
+            ReturnErrorOnFailure(encoder.Encode(selectedLocation));
+        }
+        return CHIP_NO_ERROR;
+    });
+}
+
+CHIP_ERROR Instance::ReadProgress(AttributeValueEncoder & aEncoder)
+{
+    if (mDelegate->GetNumberOfProgressElements() == 0)
+    {
+        return aEncoder.EncodeNull();
+    }
+
+    return aEncoder.EncodeList([this](const auto & encoder) -> CHIP_ERROR {
+        uint32_t locationIndex = 0;
+        Structs::ProgressStruct::Type progressElement;
+
+        while (mDelegate->GetProgressElementByIndex(locationIndex++, progressElement))
+        {
+            ReturnErrorOnFailure(encoder.Encode(progressElement));
+        }
+        return CHIP_NO_ERROR;
+    });
+}
+
+//*************************************************************************
+// command handlers
+
+void Instance::HandleSelectLocationsCmd(HandlerContext & ctx, const Commands::SelectLocations::DecodableType & req)
+{
+    ChipLogDetail(Zcl, "Service Area: HandleSelectLocationsCmd");
+
+    uint32_t listIndex = 0;
+    uint32_t oldSelectedLocation;
+    size_t numberOfLocations = 0;
+    bool matchesCurrentSelectedLocations;
+
+    // On receipt of this command the device SHALL respond with a SelectLocationsResponse command.
+    auto exitResponse = [ctx](SelectLocationsStatus status, CharSpan statusText) {
+        Commands::SelectLocationsResponse::Type response{
+            .status     = status,
+            .statusText = Optional(statusText),
+        };
+        ctx.mCommandHandler.AddResponse(ctx.mRequestPath, response);
+    };
+
+    // Get the number of Selected Locations in the command parameter and check that it is valid.
+    if (!req.newLocations.IsNull())
+    {
+        if (CHIP_NO_ERROR != req.newLocations.Value().ComputeSize(&numberOfLocations))
+        {
+            exitResponse(SelectLocationsStatus::kInvalidSet, "decoding of newLocations parameter size failed"_span);
+            return;
+        }
+
+        // If the device determines that it can't operate at all locations from the list,
+        // the SelectLocationsResponse command's Status field SHALL indicate InvalidSet.
+        if (((numberOfLocations == 0) || (numberOfLocations > kMaxNumSelectedLocations)))
```

Comment:
Per spec, empty list and null must have the same behavior, no?

I know that's not what the cluster spec says, but the cluster spec is violating the data model spec here.

Reply:
According to the cluster spec, receiving a list of 0 is invalid. Are you saying that this is incorrect and that we should be treating 0 as null?

Reply:
I am saying that per spec any time a list is nullable, a list of length 0 must also be allowed (and _is_ allowed, no matter what the constraint column says, so that you can define whatever constraints you want easily, and the actual behavior will also allow length 0), and length-0 list and null list must have the same semantics.

And yes, if the spec explicitly (not via constraint, which is already handled) says receiving a list of 0 length is invalid for a nullable list, that spec is wrong and needs to be fixed.

In this code, 0 length should be treated as null.

Reply:
OK. Understood. On the spec side this is already fixed since it now allows an empty list and is not nullable.

Reply:
Fixed [here](https://github.com/project-chip/connectedhomeip/pull/33991/commits/f4535dbc9a3033598208cc0b2a90390b964c61b0).